src/eval.ml

```diff
@@ -218,11 +265,90 @@ and eq_modulo : term -> term -> bool = fun a b ->
   let res = try eq_modulo [(a,b)]; true with Exit -> false in
   if !log_enabled then log_eqmd (r_or_g res "%a == %a") pp a pp b; res
 
+(** [tree_walk t d s] tries to match stack [s] against tree [t] of depth [d]. *)
+and tree_walk : Dtree.t -> int -> term list -> term option = fun tree d stk ->
+  let vars = if d > 0 then Array.make d Kind else [||] in (* dummy terms *)
+  let module R = Dtree.ReductionStack in
+  let stk = R.of_list stk in
+  (* [walk t s c] where [s] is the stack of terms to match and [c] the cursor
+     indicating where to write in the [env] array described in {!module:Terms}
+     as the environment of the RHS during matching. *)
+  let rec walk tree stk cursor =
+    match tree with
+    | Fail                                   -> None
+    | Leaf(env_builder, act)                 ->
+        (* Allocate an environement for the action. *)
+        let env = Array.make (Bindlib.mbinder_arity act) TE_None in
+        (* Retrieve terms needed in the action from the [vars] array. *)
+        let fn pos slot =
+          let t = vars.(pos) in
+          let b = Bindlib.raw_mbinder [||] [||] 0 mkfree (fun _ -> t) in
+          env.(slot) <- TE_Some(b)
+        in
+        IntMap.iter fn env_builder;
+        (* Actually perform the action. *)
+        Some(Bindlib.msubst act env)
+    | Node({swap; children; store; default}) ->
+        (* Quit if stack is too short. *)
+        if R.is_empty stk || swap >= R.length stk then None else
+        (* Pick the right term in the stack. *)
+        let (left, examined, right) = R.destruct stk swap in
+        (* Store hd of stack if needed *)
+        if store then begin
+          if !log_enabled then log_eval "(node) storing [%a]" pp examined ;
+          vars.(cursor) <- examined
+        end ;
+        let cursor = if store then succ cursor else cursor in
+        let r_ex = to_tref examined in
+        (* Fetch the right subtree and the new stack *)
+        (* [choose t] chooses a tree among {!val:children} when term [t] is
+           examined and returns the new head of stack. *)
+        let choose t : tree option * term list =
+          let h, args = get_args t in
+          let args = List.map ensure_tref args in
+          let c_ari = List.length args in
+          match h with
+          | Symb(s, _)  ->
+            r_ex := Some(to_term_n h args) ;
```

Comment:
Why not doing `r_ex := Some t`?

Reply:
Because the arguments have been change to `Tref(ref (Some t))`, for `t` initially an argument of `h` in our case.